docs/wiki/configmgmt.md
# Configuration Management

We provide a mechanism for storing and retrieving configurations of different clients, servers and services securely, any type of a resource or a client that needs configuration, it can simply implemented using base classes.

- [Introduction](#Introduction)
- [Framework configurations](#configurations)
- [Storage](#storage)
- [Encryption](#encryption)
- [More on base classes](#more-on-base-classes)


## Introduction

To define any type of resource, you can create a class, with different configuration fields, and a factory, which will manage this class instances.

Consider having remote machines, where every machine have an ip address and a region:

```python
from enum import Enum

from jumpscale.core.base import Base, fields, StoredFactory
from jumpscale.clients.sshclient import sshclient

class Region(Enum):
    US = "us"
    EU = "eu"

class Machine(Base):
    ip = fields.IPAddress()
    region = fields.Enum(Region)

machines = StoredFactory(Machine)
```

This simply define a resource called `Machine`, which have two fields.

`machines` here is factory, and it's the entry point where we can create and manage instances of this machines:

```python
m1 = machines.get("m1")
m1.ip = "127.0.0.1"
m1.region = "us"
m1.save()
```

`save` will store this configuration to current configured storage, and at any time, you can list current machines, delete any of them...etc:

```python
print(machines.list_all())
machines.delete("m1")
```

## configurations

"""Config module is the single entry of configurations across the framework.

It allows
- resolving configurations paths for configuration directory`config_root`, or configuration file path `config_path`
- rebuildling default configurations or retrieving them (using `get_default_config`)
- Getting configurations using `get_default_config`
- Updating configuration using `update_config`



JS-NG> j.core.config
<module 'jumpscale.core.config' from '/home/ahmed/wspace/js/js-ng/jumpscale/core/config/__init__.py'>

### Getting where is the config.toml path
```
JS-NG> j.core.config.config_path
'/home/ahmed/.config/jumpscale/config.toml'
```

### Getting configuration directory
```
JS-NG> j.core.config.config_root
'/home/ahmed/.config/jumpscale'
```

### Getting default configurations of js-ng (used in case of no configuration found)
```
JS-NG> j.core.config.get_default_config()
{'debug': True, 'logging': {'redis': {'enabled': True, 'level': 15, 'max_size': 1000, 'dump': True, 'dump_dir': '/home/ahmed/.config/jumpscale/logs/redis'}, 'filesystem': {'enabled': True, 'level': 15, 'log_dir': '/home/ahmed/.config/jumpscale/logs/fs/log.txt', 'rotation': '5 MB'}}, 'alerts': {'enabled': True, 'level': 40}, 'ssh_key_path': '', 'private_key_path': '', 'stores': {'redis': {'hostname': 'localhost', 'port': 6379}, 'filesystem': {'path': '/home/ahmed/.config/jumpscale/secureconfig'}}, 'store': 'filesystem', 'threebot': {'default': ''}, 'explorer': {'default_url': 'https://explorer.testnet.grid.tf/explorer'}}
```

### Getting the current configurations
```
JS-NG> j.core.config.get_config()
{'debug': True, 'ssh_key_path': '', 'log_to_redis': False, 'log_to_files': True, 'log_level': 15, 'private_key_path': '/home/ahmed/.config/jumpscale/mykey.priv', 'secure_config_path': '/home/ahmed/.config/jumpscale/secureconfig', 'store': 'filesystem', 'logging': {'handlers': [{'sink': 'sys.stdout', 'format': '{time} - {message}', 'colorize': True, 'enqueue': True}, {'sink': '/home/ahmed/.config/jumpscale/logs/file_jumpscale.log', 'serialize': True, 'enqueue': True}], 'redis': {'enabled': True, 'level': 15, 'max_size': 1000, 'dump': True, 'dump_dir': '/home/ahmed/.config/jumpscale/logs/redis'}, 'filesystem': {'enabled': True, 'level': 15, 'log_dir': '/home/ahmed/.config/jumpscale/logs/fs/log.txt', 'rotation': '5 MB'}}, 'stores': {'redis': {'hostname': 'localhost', 'port': 6379}, 'filesystem': {'path': '/home/ahmed/.config/jumpscale/secureconfig'}}, 'alerts': {'enabled': True, 'level': 40}, 'threebot': {'default': ''}, 'explorer': {'default_url': 'https://explorer.testnet.grid.tf/explorer'}}
```

### Updating configurations

```
JS-NG> conf = j.core.config.get_default_config()
JS-NG> conf
{'debug': True, 'ssh_key_path': '', 'log_to_redis': False, 'log_to_files': True, 'log_level': 15, 'private_key_path': '/home/ahmed/.config/jumpscale/mykey.priv', 'secure_config_path': '/home/ahmed/.config/jumpscale/secureconfig', 'store': 'filesystem', 'logging': {'handlers': [{'sink': 'sys.stdout', 'format': '{time} - {message}', 'colorize': True, 'enqueue': True}, {'sink': '/home/ahmed/.config/jumpscale/logs/file_jumpscale.log', 'serialize': True, 'enqueue': True}], 'redis': {'enabled': True, 'level': 15, 'max_size': 1000, 'dump': True, 'dump_dir': '/home/ahmed/.config/jumpscale/logs/redis'}, 'filesystem': {'enabled': True, 'level': 15, 'log_dir': '/home/ahmed/.config/jumpscale/logs/fs/log.txt', 'rotation': '5 MB'}}, 'stores': {'redis': {'hostname': 'localhost', 'port': 6379}, 'filesystem': {'path': '/home/ahmed/.config/jumpscale/secureconfig'}}, 'alerts': {'enabled': True, 'level': 40}, 'threebot': {'default': ''}, 'explorer': {'default_url': 'https://explorer.testnet.grid.tf/explorer'}}

JS-NG> conf['favcolor'] = 'blue'
JS-NG> j.core.config.update_config(conf)
JS-NG> j.core.config.get_config()
{'debug': True, 'ssh_key_path': '', 'log_to_redis': False, 'log_to_files': True, 'log_level': 15, 'private_key_path': '/home/ahmed/.config/jumpscale/mykey.priv', 'secure_config_path': '/home/ahmed/.config/jumpscale/secureconfig', 'store': 'filesystem', 'favcolor': 'blue', 'logging': {'handlers': [{'sink': 'sys.stdout', 'format': '{time} - {message}', 'colorize': True, 'enqueue': True}, {'sink': '/home/ahmed/.config/jumpscale/logs/file_jumpscale.log', 'serialize': True, 'enqueue': True}], 'redis': {'enabled': True, 'level': 15, 'max_size': 1000, 'dump': True, 'dump_dir': '/home/ahmed/.config/jumpscale/logs/redis'}, 'filesystem': {'enabled': True, 'level': 15, 'log_dir': '/home/ahmed/.config/jumpscale/logs/fs/log.txt', 'rotation': '5 MB'}}, 'stores': {'redis': {'hostname': 'localhost', 'port': 6379}, 'filesystem': {'path': '/home/ahmed/.config/jumpscale/secureconfig'}}, 'alerts': {'enabled': True, 'level': 40}, 'threebot': {'default': ''}, 'explorer': {'default_url': 'https://explorer.testnet.grid.tf/explorer'}}
```

### Get/Set
you can use `j.core.config.get` and `j.core.config.set` to retrive values of keys or set the value of a key respectively.

e.g

```
JS-NG> j.core.config.get("favcolor")
'blue'

JS-NG> j.core.config.set("favcolor", "grey")
JS-NG> j.core.config.get("favcolor")
'grey'
```




## Storage

Any stored factory has a store, which is selected based configuration.

For now, we have two types of storage:

* File system: by default, stores secure config under `~/.config/jumpscale/secureconfig`
* Redis: stores configuration at redis.


Implementation can be found at `jumpscale.core.base.store`, the main idea is that, every factory should store configuration on a unique location, this location is automatically generated, and instances are stored in a tree-like hierarchy.

For example:

```
➜  js-ng git:(development) ✗ tree ~/.config/jumpscale/secureconfig
/home/abom/.config/jumpscale/secureconfig
├── jumpscale
│   ├── clients
│   │   ├── docker
│   │   │   └── docker
│   │   │       └── DockerClient
│   │   │           └── test
│   │   │               └── data
│   │   ├── gdrive
│   │   │   └── gdrive
│   │   │       └── GdriveClient
│   │   │           └── test
│   │   │               └── data
│   │   ├── github
│   │   │   └── github
│   │   │       ├── Github
│   │   │       │   ├── g1
│   │   │       │   │   ├── data
│   │   │       │   │   └── users
│   │   │       │   │       └── jumpscale
│   │   │       │   │           └── clients
│   │   │       │   │               └── github
│   │   │       │   │                   └── github
│   │   │       │   │                       └── User
│   │   │       │   │                           └── u1
│   │   │       │   │                               └── data
```

### Serialization:

Only `json` serialization is now supported, if you need to get a serializable dict for any base class instance, just use `to_dict()` method.


## Encryption

We use public key encryption, implementation is at `j.data.nacl`.
For keys, we use the key that's generated and saved at `private_key_path`.
There is a key that's auto-generated for you, to generate a new key, use `jsctl` to generate a new key and update the config:

```
jsctl genkey . test
jsctl config update private_key_path "`pwd`/test.priv"
```

Only secret fields are encrypted by default, so, if you have a password field for example, it's better to use `fields.Secret` for it:

```python
class Region(enum):
    US = "us"
    EU = "eu"

class Machine(Base):
    ip = fields.IPAddress()
    region = fields.Enum(Region)
    username = fields.Secret()
    password = fields.Secret()
```

Now when the store writes any instance configuration of `Machine`, it will encrypt it.
If you load this instance again, it will be decrypted.

## More on base classes

You can check what base classes can provide, what are supported field types  and more at [Base classes](baseclasses.md).


docs/tests/core/config/test_config.md
### test_01_get_default_config

Test case for getting jsng default config.

**Test Scenario**

- Get the default config.
- Check that some of these config are having the default values.

### test_02_get_config

Test case for getting jsng config.

**Test Scenario**

- Get the config.
- Check that some of these config keys are exist.

### test_03_update_config

Test case for updating jsng config.

**Test Scenario**

- Set a new key with random value.
- Get the value this key.
- Check that the value of this key is the same has been set.
- Get the config.
- Remove this key from the config and update the config.

### test_04_migrate_config

Test case for migrating config.

**Test Scenario**

- Get the config.
- Remove factory key.
- Migrate the config.
- Get the new config.
- Update the config to the original one.
- Check that the factory key is in the new config.

### test_05_generate_key

Test case for generating key.

**Test Scenario**

- Create a temporary directory.
- Generate a key in the this directory.
- Check that the key has been generated.
- Remove this directory.

### test_06_get_environment_config_0_logging_config

Test case for getting environment config [with config_name='logging_config'].

**Test Scenario**

- Get the config.
- Get environment config.
- Check that the environment config is the same as the one used in config.

### test_06_get_environment_config_1_private_key

Test case for getting environment config [with config_name='private_key'].

**Test Scenario**

- Get the config.
- Get environment config.
- Check that the environment config is the same as the one used in config.

### test_06_get_environment_config_2_private_key_path

Test case for getting environment config [with config_name='private_key_path'].

**Test Scenario**

- Get the config.
- Get environment config.
- Check that the environment config is the same as the one used in config.

### test_06_get_environment_config_3_threebot_data

Test case for getting environment config [with config_name='threebot_data'].

**Test Scenario**

- Get the config.
- Get environment config.
- Check that the environment config is the same as the one used in config.

### test_07_get_store_config

Test case for getting store config.

**Test Scenario**

- Get the config.
- Get store config.
- Check that the store config is the same as the one used in config.
- Try to get store config for random store name, should raise error.


jumpscale/core/config/config.py
"""Config module is the single entry of configurations across the framework.

It allows
- resolving configurations paths for configuration directory`config_root`, or configuration file path `config_path`
- rebuildling default configurations or retrieving them (using `get_default_config`)
- Getting configurations using `get_default_config`
- Updating configuration using `update_config`



JS-NG> j.core.config
<module 'jumpscale.core.config' from '/home/ahmed/wspace/js/js-ng/jumpscale/core/config/__init__.py'>

## Getting where is the config.toml path
```
JS-NG> j.core.config.config_path
'/home/ahmed/.config/jumpscale/config.toml'
```

## Getting configuration directory
```
JS-NG> j.core.config.config_root
'/home/ahmed/.config/jumpscale'
```

## Getting default configurations of js-ng (used in case of no configuration found)
```
JS-NG> j.core.config.get_default_config()
{'debug': True, 'logging': {'redis': {'enabled': True, 'level': 15, 'max_size': 1000, 'dump': True, 'dump_dir': '/home/ahmed/.config/jumpscale/logs/redis'}, 'filesystem': {'enabled': True, 'level': 15, 'log_dir': '/home/ahmed/.config/jumpscale/logs/fs/log.txt', 'rotation': '5 MB'}}, 'alerts': {'enabled': True, 'level': 40}, 'ssh_key_path': '', 'private_key_path': '', 'stores': {'redis': {'hostname': 'localhost', 'port': 6379}, 'filesystem': {'path': '/home/ahmed/.config/jumpscale/secureconfig'}}, 'store': 'filesystem', 'threebot': {'default': ''}, 'explorer': {'default_url': 'https://explorer.testnet.grid.tf/explorer'}}
```

## Getting the current configurations
```
JS-NG> j.core.config.get_config()
{'debug': True, 'ssh_key_path': '', 'log_to_redis': False, 'log_to_files': True, 'log_level': 15, 'private_key_path': '/home/ahmed/.config/jumpscale/mykey.priv', 'secure_config_path': '/home/ahmed/.config/jumpscale/secureconfig', 'store': 'filesystem', 'logging': {'handlers': [{'sink': 'sys.stdout', 'format': '{time} - {message}', 'colorize': True, 'enqueue': True}, {'sink': '/home/ahmed/.config/jumpscale/logs/file_jumpscale.log', 'serialize': True, 'enqueue': True}], 'redis': {'enabled': True, 'level': 15, 'max_size': 1000, 'dump': True, 'dump_dir': '/home/ahmed/.config/jumpscale/logs/redis'}, 'filesystem': {'enabled': True, 'level': 15, 'log_dir': '/home/ahmed/.config/jumpscale/logs/fs/log.txt', 'rotation': '5 MB'}}, 'stores': {'redis': {'hostname': 'localhost', 'port': 6379}, 'filesystem': {'path': '/home/ahmed/.config/jumpscale/secureconfig'}}, 'alerts': {'enabled': True, 'level': 40}, 'threebot': {'default': ''}, 'explorer': {'default_url': 'https://explorer.testnet.grid.tf/explorer'}}
```

## Updating configurations

```
JS-NG> conf = j.core.config.get_default_config()
JS-NG> conf
{'debug': True, 'ssh_key_path': '', 'log_to_redis': False, 'log_to_files': True, 'log_level': 15, 'private_key_path': '/home/ahmed/.config/jumpscale/mykey.priv', 'secure_config_path': '/home/ahmed/.config/jumpscale/secureconfig', 'store': 'filesystem', 'logging': {'handlers': [{'sink': 'sys.stdout', 'format': '{time} - {message}', 'colorize': True, 'enqueue': True}, {'sink': '/home/ahmed/.config/jumpscale/logs/file_jumpscale.log', 'serialize': True, 'enqueue': True}], 'redis': {'enabled': True, 'level': 15, 'max_size': 1000, 'dump': True, 'dump_dir': '/home/ahmed/.config/jumpscale/logs/redis'}, 'filesystem': {'enabled': True, 'level': 15, 'log_dir': '/home/ahmed/.config/jumpscale/logs/fs/log.txt', 'rotation': '5 MB'}}, 'stores': {'redis': {'hostname': 'localhost', 'port': 6379}, 'filesystem': {'path': '/home/ahmed/.config/jumpscale/secureconfig'}}, 'alerts': {'enabled': True, 'level': 40}, 'threebot': {'default': ''}, 'explorer': {'default_url': 'https://explorer.testnet.grid.tf/explorer'}}

JS-NG> conf['favcolor'] = 'blue'
JS-NG> j.core.config.update_config(conf)
JS-NG> j.core.config.get_config()
{'debug': True, 'ssh_key_path': '', 'log_to_redis': False, 'log_to_files': True, 'log_level': 15, 'private_key_path': '/home/ahmed/.config/jumpscale/mykey.priv', 'secure_config_path': '/home/ahmed/.config/jumpscale/secureconfig', 'store': 'filesystem', 'favcolor': 'blue', 'logging': {'handlers': [{'sink': 'sys.stdout', 'format': '{time} - {message}', 'colorize': True, 'enqueue': True}, {'sink': '/home/ahmed/.config/jumpscale/logs/file_jumpscale.log', 'serialize': True, 'enqueue': True}], 'redis': {'enabled': True, 'level': 15, 'max_size': 1000, 'dump': True, 'dump_dir': '/home/ahmed/.config/jumpscale/logs/redis'}, 'filesystem': {'enabled': True, 'level': 15, 'log_dir': '/home/ahmed/.config/jumpscale/logs/fs/log.txt', 'rotation': '5 MB'}}, 'stores': {'redis': {'hostname': 'localhost', 'port': 6379}, 'filesystem': {'path': '/home/ahmed/.config/jumpscale/secureconfig'}}, 'alerts': {'enabled': True, 'level': 40}, 'threebot': {'default': ''}, 'explorer': {'default_url': 'https://explorer.testnet.grid.tf/explorer'}}
```

## Get/Set
you can use `j.core.config.get` and `j.core.config.set` to retrive values of keys or set the value of a key respectively.

e.g

```
JS-NG> j.core.config.get("favcolor")
'blue'

JS-NG> j.core.config.set("favcolor", "grey")
JS-NG> j.core.config.get("favcolor")
'grey'
```
"""

import os

import nacl.utils
import nacl.encoding
import pytoml as toml

try:
    from importlib import metadata
except ImportError:
    # Running on pre-3.8 Python; use importlib-metadata package
    import importlib_metadata as metadata

from nacl.public import PrivateKey


__all__ = [
    "config_path",
    "config_root",
    "get_default_config",
    "get_config",
    "update_config",
    "Environment",
    "get",
    "set",
    "set_default",
    "get_current_version",
    "generate_key",
]


package_name = "js-ng"
__version__ = metadata.version(package_name)


config_root = os.path.expanduser(os.path.join("~/.config", "jumpscale"))
config_path = os.path.join(config_root, "config.toml")


def get_current_version():
    return __version__


def get_default_config():
    """retrieves default configurations for plain jumpscale

    Returns:
        dict: default configuration
    """
    return {
        "debug": True,
        "shell": "ptpython",
        "logging": {
            "redis": {
                "enabled": True,
                "level": 15,
                "max_size": 1000,
                "dump": True,
                "dump_dir": os.path.join(config_root, "logs/redis"),
            },
            "filesystem": {
                "enabled": True,
                "level": 15,
                "log_dir": os.path.join(config_root, "logs/fs/log.txt"),
                "rotation": "5 MB",
            },
        },
        "alerts": {"enabled": True, "level": 40},
        "ssh_key_path": "",
        "private_key_path": "",
        "stores": {
            "redis": {"hostname": "localhost", "port": 6379},
            "mongo": {"hostname": "localhost", "port": 27017},
            "filesystem": {"path": os.path.expanduser(os.path.join(config_root, "secureconfig"))},
            "whoosh": {"path": os.path.expanduser(os.path.join(config_root, "whoosh_indexes"))},
        },
        "factory": {"always_reload": False},
        "store": "filesystem",
        "threebot": {"default": "",},
        "explorer": {"default_url": "https://explorer.testnet.grid.tf/explorer",},
    }


def get_config():
    """Gets jumpscale configurations

    Returns:
        [dict] - toml loaded config of CONFIG_DIR/config.toml
    """
    with open(config_path, "r") as f:
        return toml.load(f)


def update_config(data):
    """Update jumpscale config with new data

    Arguments:
        data {dict} -- dict to update the config with.
    """
    with open(config_path, "w") as f:
        toml.dump(data, f)


def get(key, default=None):
    """ Retrives value from jumpscale config

    Arguments:
        key (str): the key you wish to retrieve
        default (object): return value if key doesn't exist in configurations
    """
    conf = get_config()
    return conf.get(key, default)


def set(key, val):
    """ Sets value in jumpscale config

    Arguments:
        key (str): the key you wish to update
        val: value to update with
    """
    conf = get_config()
    conf[key] = val
    update_config(conf)


def set_default(key, val):
    """ Sets key to value in jumpscale config and returns

    Arguments:
        key (str): the key you wish to update
        val: value to update with and to return if key doesn't exist in configurations

    Returns:
        val (str): returned if key doesn't exist in configuration or the value of key in configurations
    """
    conf = get_config()
    if key in conf:
        return conf[key]
    else:
        set(key, val)
        return val


def migrate_config():
    """add missing keys to current config from default"""

    default_config = get_default_config()
    current_config = get_config()

    def copy_missing(default, current):
        for key, value in default.items():
            if key not in current:
                current[key] = value
            elif isinstance(value, dict):
                copy_missing(value, current[key])

    copy_missing(default_config, current_config)
    update_config(current_config)


# Create default configurations file on loading.
if not os.path.exists(config_path):
    os.makedirs(os.path.dirname(config_path), exist_ok=True)
    with open(config_path, "w") as f:
        # don't ues mknod https://github.com/threefoldtech/js-ng/issues/182
        f.close()


def generate_key(basepath):
    hsk = PrivateKey.generate()
    hpk = hsk.public_key
    skpath = basepath + ".priv"
    pkpath = basepath + ".pub"
    with open(skpath, "wb") as f:
        f.write(hsk.encode(nacl.encoding.Base64Encoder()))
    with open(pkpath, "wb") as f:
        f.write(hpk.encode(nacl.encoding.Base64Encoder()))
    return skpath


class StoreTypeNotFound(Exception):
    pass


class Environment:
    def get_private_key_path(self):
        config = get_config()
        private_key_path = config["private_key_path"]
        return private_key_path

    def get_threebot_data(self):
        config = get_config()
        return config.get("threebot", {})

    def get_private_key(self):
        private_key_path = self.get_private_key_path()
        if not private_key_path:
            keypath = os.path.join(config_root, "privatekey")
            private_key_path = generate_key(keypath)
            config = get_config()
            config["private_key_path"] = private_key_path
            update_config(config)
        return open(private_key_path, "rb").read()

    def get_store_config(self, name):
        config = get_config()
        stores = config["stores"]
        if name not in stores:
            raise StoreTypeNotFound(f"'{name}' store is not found")
        return stores[name]

    def get_logging_config(self):
        return get_config()["logging"]


migrate_config()
